### PR TITLE
Event filtering

### PR DESCRIPTION
**Pull request by reed-atakama**
Easier to read diff in split view

### CODE REVIEW

File: cloudsync_gdrive.py [hunk 1]
```diff
@@ -312,56 +326,103 @@ def events(self) -> Generator[Event, None, None]:  # pylint: disable=too-many-lo
...
+
+        if not event:
+            return EventFilter.IGNORE
+
+        if not self._root_path:
+            return EventFilter.PROCESS
+
+        state_path = self.sync_state.get_path(event.oid)
+        prior_subpath = self.is_subpath_of_root(state_path)
+        if not event.exists:
```
Comment (howufiger7) [inline cloudsync_gdrive.py:400]: curious - are gdrive delete events the same as onedrive in that they don't have a path?

File: cloudsync_gdrive.py [hunk 2]
```diff
@@ -312,56 +326,103 @@ def events(self) -> Generator[Event, None, None]:  # pylint: disable=too-many-lo
...
+            curr_subpath = self.is_subpath_of_root(event.path)
+            if curr_subpath and not prior_subpath:
+                # Can't differentiate between creates and renames without watching the entire filesystem:
+                # Event has an oid and a current path, its a rename if the oid was seen before,
+                # but since events outside root are ignored we don't catch the case where an item is
+                # created outside root and then renamed into root.
+                # Hence the walk for directories -- a tradeoff for ignoring "outside root" events.
+                log.debug("created in or renamed into root: %s", event.path)
+                if event.otype == DIRECTORY:
+                    return EventFilter.WALK
```
Comment (howufiger7) [inline cloudsync_gdrive.py:414]: the main reason we needed this for OneDrive is that OneDrive generates a single event for folder renames (the event is for the folder being renamed, and no events come in for child files/folders)

does gdrive behave the same way?

File: cloudsync_gdrive.py [hunk 3]
```diff
@@ -312,56 +326,103 @@ def events(self) -> Generator[Event, None, None]:  # pylint: disable=too-many-lo
...
-                path = self._path_oid(oid, use_cache=False)
+            if new_cursor and page_token and new_cursor != page_token:
+                self.__cursor = new_cursor
+            page_token = response.get('nextPageToken')
 
-                event = Event(otype, oid, path, ohash, exists, ts, new_cursor=new_cursor)
+    def _convert_to_event(self, change, new_cursor):
+        ts = arrow.get(change.get('time')).float_timestamp
+        oid = change.get('fileId')
+        exists = None
```
Comment (howufiger7) [inline cloudsync_gdrive.py:351]: what does event.exists = None mean?

Comment (reed-atakama) [reply cloudsync_gdrive.py:351]: This means we don't know if it exists or doesn't exist and that should be determined by `get_latest`. This needs to be supported and is related to the cloudsync PR Erik put up yesterday https://github.com/AtakamaLLC/cloudsync/pull/329/files. For gdrive specifically, we can avoid the an unnecessary API hit for each event if we leave exists as None and wait for get_latest to populate this field

File: cloudsync_gdrive.py (hunk 1 above)
Comment (reed-atakama) [reply cloudsync_gdrive.py:400]: The current implementation populates a path but we are moving away from that. Path will be `None` for all events going forward to cut down on duplicate api hits

File: cloudsync_gdrive.py (hunk 2 above)
Comment (reed-atakama) [reply cloudsync_gdrive.py:414]: Yes, just double checked, gdrive behaves the sames way.

**Review by howufiger7 — APPROVED** [review]

**Pull request merged**